src/mnode/mnode-pastel.cpp

```diff
@@ -1442,3 +1442,132 @@ bool CTakeDownTicket::FindTicketInDb(const std::string& key, CTakeDownTicket& ti
+    {
+        LOCK(cs_main);
+        chainHeight = static_cast<unsigned int>(chainActive.Height()) + 1;
+    }
+
+    // A. Check if the username is already registered on the blockchain.
+    if (masterNodeCtrl.masternodeTickets.CheckTicketExist(*this)) {
+        throw std::runtime_error(strprintf("This Username is already registered in blockchain [Username = %s]", username));
+    }
+
+    // B Verify signature
+    // We will check that it is the correct PastelID
+    std::string strThisTicket = ToStr();
+    if (!CPastelID::Verify(reinterpret_cast<const unsigned char *>(strThisTicket.c_str()), strThisTicket.size(),
+                           signature.data(), signature.size(),
+                           pastelID))
+    {
+        throw std::runtime_error(strprintf("%s ticket's signature is invalid. PastelID - [%s]", GetTicketDescription(TicketID::Username), pastelID));
+    }
+    
+    // C (ticket transaction replay attack protection)
+    CChangeUsernameTicket existingTicket;
+    if ((FindTicketInDb(username, existingTicket)) &&
+        (!existingTicket.IsBlock(m_nBlock) || existingTicket.m_txid != m_txid))
+    {
+        throw std::runtime_error(strprintf("This Username Change Request is already registered in blockchain [Username = %s]"
+                           "[this ticket block = %u txid = %s; found ticket block  = %u txid = %s]",
+                           username, existingTicket.GetBlock(), existingTicket.m_txid, m_nBlock, m_txid));
+    }
+
+    // D. Check if this PastelID hasn't changed Username in last 24 hours.
+    CChangeUsernameTicket _ticket;
+    _ticket.pastelID = pastelID;
+    if (masterNodeCtrl.masternodeTickets.FindTicketBySecondaryKey(_ticket)) {
+        const unsigned int height = (preReg || IsBlock(0)) ? chainHeight : m_nBlock;
+        if (height - _ticket.m_nBlock <= 24 * 24) {
+            // D.2 IF PastelID has changed Username in last 24 hours (~24*24 blocks), do not allow them to change
+            throw std::runtime_error(strprintf("%s ticket is invalid. Already changed in last 24 hours. PastelID - [%s]", GetTicketDescription(TicketID::Username), pastelID));
+        }
+    }
+
+    // E. Check if address has coins to pay for Username Change Ticket
+    if (preReg) 
+    {
+        if (!masterNodeCtrl.masternodeTickets.FindTicketBySecondaryKey(_ticket)) {
+            if (fee == 5000) {
+                throw std::runtime_error(strprintf("%s ticket's fee is invalid. PastelID - [%s], invalid fee - [%d]", GetTicketDescription(TicketID::Username), pastelID, fee));
+            }
+        } else {
+            if (fee == 100) {
+                throw std::runtime_error(strprintf("%s ticket's fee is invalid. PastelID - [%s], invalid fee - [%d]", GetTicketDescription(TicketID::Username), pastelID, fee));
+            }
+        }
+        const auto fullTicketPrice = TicketPrice(chainHeight);
+        
+        if (pwalletMain->GetBalance() < fullTicketPrice * COIN) 
```

Comment:
only this part for preReg, because it uses current balance
these two checks above for 5K and 100  should be for any run of IsValid, also can you do 5K and 100 as parameters in mnode-controller.h/cpp, pelase

Reply:
regarding to your statement "also can you do 5K and 100 as parameters", I understand that we add 2 parameters: firstChangeFree, changeAgainFee. Does my understand correct?

Reply:
fixed